Library/OcConsoleLib/TextOutputBuiltin.c

```diff
@@ -620,6 +82,89 @@ STATIC EFI_CONSOLE_CONTROL_SCREEN_MODE      mConsoleMode = EfiConsoleControlScre
 #define MIN_SUPPORTED_CONSOLE_WIDTH   (80)
 #define MIN_SUPPORTED_CONSOLE_HEIGHT  (25)
 
+BOOLEAN
+GetConsoleFontCharInfo (
+  IN  OC_CONSOLE_FONT       *ConsoleFont,
+  IN  CHAR16                Char,
+  OUT OC_CONSOLE_FONT_PAGE  **Page,
+  OUT UINT8                 *GlyphIndex,
+  IN  BOOLEAN               UseFallback
+  )
+{
+  BOOLEAN  Success;
+  UINT16   PageNumber;
+  UINT8    PageChar;
+  UINT16   PageIndex;
+
+  ASSERT (Page != NULL);
+  ASSERT (GlyphIndex != NULL);
+
+  PageNumber = (Char >> 7) & 0x1FF;
+  PageChar   = Char & 0x7F;
+
+  *GlyphIndex = 0;
+
+  Success = TRUE;
+  do {
+    if ((PageNumber >= ConsoleFont->PageMin) && (PageNumber < ConsoleFont->PageMax)) {
+      PageIndex = PageNumber - ConsoleFont->PageMin + 1;
+      if (ConsoleFont->PageOffsets != NULL) {
+        PageIndex = ConsoleFont->PageOffsets[PageIndex - 1];
+      }
+
+      if (PageIndex != 0) {
+        *Page = &ConsoleFont->Pages[PageIndex - 1];
+        if ((PageChar >= (*Page)->CharMin) && (PageChar < (*Page)->CharMax)) {
+          ASSERT ((*Page)->Glyphs != NULL);
+          if ((*Page)->GlyphOffsets == NULL) {
+            *GlyphIndex = PageChar - (*Page)->CharMin + 1;
+          } else {
+            *GlyphIndex = (*Page)->GlyphOffsets[PageChar - (*Page)->CharMin];
+          }
+        }
+      }
+    }
+
+    if (*GlyphIndex == 0) {
+      //
+      // Give up and render nothing if fallback char is not present.
+      //
+      ASSERT (Success == TRUE);
+      if (!UseFallback || (Success == FALSE)) {
+        return FALSE;
+      }
+
+      Success = FALSE;
+
+      //
+      // For default font, render all unknown chars in first two pages as space (following the explicit
+      // spaces in free console font used in XNU); in other pages and in user fonts use underscore.
+      //
+      if (mIsDefaultFont && (PageNumber < 2)) {
```

Comment:
Note to self: If I'm not including ISO Latin page 1 - as I'm currently not in this PR - this PageNumber test should be 1, not 2. To change, if we are keeping the special case code.